lib/galaxy/metadata/__init__.py

```diff
@@ -0,0 +1,277 @@
+                                    max_metadata_value_size)
+        if include_command:
+            # return command required to build
+            fd, fp = tempfile.mkstemp(suffix='.py', dir=tmp_dir, prefix="set_metadata_")
+            metadata_script_file = abspath(fp)
+            os.fdopen(fd, 'w').write('from galaxy_ext.metadata.set_metadata import set_metadata; set_metadata()')
+            return 'python "%s" %s' % (metadata_path_on_compute(metadata_script_file), args)
+        else:
+            # return args to galaxy_ext.metadata.set_metadata required to build
+            return args
+
+    def external_metadata_set_successfully(self, dataset, sa_session):
+        metadata_files = self._get_output_filenames_by_dataset(dataset, sa_session)
+        if not metadata_files:
+            return False  # this file doesn't exist
+        rval, rstring = json.load(open(metadata_files.filename_results_code))
+        if not rval:
+            log.debug('setting metadata externally failed for %s %s: %s' % (dataset.__class__.__name__, dataset.id, rstring))
+        return rval
+
+    def cleanup_external_metadata(self, sa_session):
+        log.debug('Cleaning up external metadata files')
+        for metadata_files in sa_session.query(galaxy.model.Job).get(self.job_id).external_output_metadata:
+            # we need to confirm that any MetadataTempFile files were removed, if not we need to remove them
+            # can occur if the job was stopped before completion, but a MetadataTempFile is used in the set_meta
+            MetadataTempFile.cleanup_from_JSON_dict_filename(metadata_files.filename_out)
+            dataset_key = self._get_dataset_metadata_key(metadata_files.dataset)
+            for key, fname in [('filename_in', metadata_files.filename_in),
+                               ('filename_out', metadata_files.filename_out),
+                               ('filename_results_code', metadata_files.filename_results_code),
+                               ('filename_kwds', metadata_files.filename_kwds),
+                               ('filename_override_metadata', metadata_files.filename_override_metadata)]:
+                try:
+                    os.remove(fname)
+                except Exception as e:
+                    log.debug('Failed to cleanup external metadata file (%s) for %s: %s' % (key, dataset_key, e))
+
+    def set_job_runner_external_pid(self, pid, sa_session):
+        for metadata_files in sa_session.query(galaxy.model.Job).get(self.job_id).external_output_metadata:
+            metadata_files.job_runner_external_pid = pid
+            sa_session.add(metadata_files)
+            sa_session.flush()
+
+    def load_metadata(self, dataset, dataset_name, sa_session, working_directory, remote_metadata_directory=None):
+        # load metadata from file
+        # we need to no longer allow metadata to be edited while the job is still running,
```

Comment:
off-topic, but we should fix this (is that even enforced anywhere?). Since we record HDA versions now this shouldn't be an issue anymore.

Reply:
Right, Im not sure why I thought this wouldn't be an issue anymore.